docs/codeql/codeql-language-guides/analyzing-data-flow-in-go.rst

```diff
@@ -0,0 +1,369 @@
+
+   import go
+
+   from Function osOpen, CallExpr call
+   where
+     osOpen.hasQualifiedName("os", "Open") and
+     call.getTarget() = osOpen
+   select call.getArgument(0)
+
+Unfortunately, this only gives the expression in the argument, not the values which could be passed to it. So we use local data flow to find all expressions that flow into the argument:
+
+.. code-block:: ql
+
+   import go
+
+   from Function osOpen, CallExpr call, Expr src
+   where
+     osOpen.hasQualifiedName("os", "Open") and
+     call.getTarget() = osOpen and
+     DataFlow::localFlow(DataFlow::exprNode(src), DataFlow::exprNode(call.getArgument(0)))
+   select src
+
+Then we can make the source more specific, for example an access to a parameter. This query finds where a public parameter is passed to ``os.Open(..)``:
+
+.. code-block:: ql
+
+   import go
+
+   from Function osOpen, CallExpr call, Parameter p
+   where
+     osOpen.hasQualifiedName("os", "Open") and
+     call.getTarget() = osOpen and
+     DataFlow::localFlow(DataFlow::parameterNode(p), DataFlow::exprNode(call.getArgument(0)))
+   select p
+
+This query finds calls to formatting functions where the format string is not hard-coded.
+
+.. code-block:: ql
+
+   import go
+
+   from StringOps::Formatting::Range format, CallExpr call, Expr formatString
+   where
+     call.getTarget() = format and
+     formatString = call.getArgument(format.getFormatStringIndex()) and
+     not exists(DataFlow::Node source, DataFlow::Node sink |
+       DataFlow::localFlow(source, sink) and
+       source.asExpr() instanceof StringLit and
+       sink.asExpr() = formatString
+     )
+   select call, "Argument to String format method isn't hard-coded."
+
+Exercises
+~~~~~~~~~
+
+Exercise 1: Write a query that finds all hard-coded strings used to create a ``url.URL``, using local data flow. (`Answer <#exercise-1>`__)
+
+Global data flow
+----------------
+
+Global data flow tracks data flow throughout the entire program, and is therefore more powerful than local data flow. However, global data flow is less precise than local data flow, and the analysis typically requires significantly more time and memory to perform.
+
+.. pull-quote:: Note
+
+   .. include:: ../reusables/path-problem.rst
+
+Using global data flow
+~~~~~~~~~~~~~~~~~~~~~~
+
+The global data flow library is used by implementing the signature ``DataFlow::ConfigSig`` and applying the module ``DataFlow::Global<ConfigSig>``:
+
+.. code-block:: ql
+
+   import go
+
+   module MyFlowConfiguration implements DataFlow::ConfigSig {
+     predicate isSource(DataFlow::Node source) {
+       ...
+     }
+
+     predicate isSink(DataFlow::Node sink) {
+       ...
+     }
+   }
+
+   module MyFlow = DataFlow::Global<MyFlowConfiguration>;
+
+These predicates are defined in the configuration:
+
+-  ``isSource`` - defines where data may flow from.
+-  ``isSink`` - defines where data may flow to.
+-  ``isBarrier`` - optional, restricts the data flow.
```

Comment:
Is there a better description than "restricts"? All the predicates "restrict" data flow in some way.
```suggestion
-  ``isBarrier`` - optional, breaks the data flow.
```

Reply:
I've changed this to `optional, defines where data flow is blocked` in all the language guides where it appears.